File: LeetCode/5.longest-palindromic-substring.java
/*
 * @lc app=leetcode id=5 lang=java
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
class Solution {
    public String longestPalindrome(String s) {
        String longest = "";
        for (int i = 0; i < s.length(); i++) {
            String curr = this.longestPalindormeWhereXisItsCenter(s, i);
            if (curr.length() > longest.length()) {
                longest = curr;
            }
        }
        return longest;
    }

    static String longestPalindormeWhereXisItsCenter(String s, Integer idx) {
        Integer r = idx;
        Integer l = idx;

        if (s.length() % 2 == 0) {
            r = (s.length() / 2) - 1;
            l = (s.length() / 2);
        }

        Integer c_l = l;
        Integer c_r = r;

        if (s.length() % 2 == 0) {
            while (l > -1 || r < s.length()) {
                if (s.charAt(l) == s.charAt(r)) {
                    c_l = l;
                    c_r = r;
                    l--;
                    r++;
                } else {
                    
                }
            }
        }
        if (s.length() % 2 != 0) {
            while (l > -1 && r < s.length()) {

            }
        }

        return s.substring(c_l, c_r + 1);
    }
}
// @lc code=end

File: LeetCode/5.longest-palindromic-substring.py
#
# @lc app=leetcode id=5 lang=python3
#
# [5] Longest Palindromic Substring
#

# @lc code=start
class Solution:
    def longestPalindrome(self, s: str) -> str:
        s1 = s
        s2 = s[::-1]

        length = len(s) + 1

        dp = [[0 for _ in range(length) ] for _ in range(length)]

        maxVal = 0
        position = (0, 0)
        for i in range(1, length):
            for j in range(1, length):
                if s1[i-1] == s2[j-1]:
                    dp[i][j] = dp[i-1][j-1] + 1
                    if dp[i][j] > maxVal:
                        maxVal = dp[i][j]
                        position = (i -1, j-1)
        return s[position[0] - maxVal + 1 : position[0] + 1]

        
        
# @lc code=end

s= Solution()
print(s.longestPalindrome("babad"))
